Object Tutorials Chapter 4 (#1848)
MystenLabs/sui: doc/src/build/programming-with-objects/ch4-object-wrapping.md

## Chapter 4: Object Wrapping
In most languages, we organize data structures in layer by nesting complex data structures in another data structure. In Move, you could do the same by putting a field of struct type in another, like the following:
```rust
struct Foo has key {
    id: VersionedID,
    bar: Bar,
}
struct Bar has store {
    value: u64,
}
```
> :bulb: For a struct type to be capable of being embedded in a Sui object struct (which will have `key` ability), the embedded struct type must have `store` ability.

> In the above example, `Bar` is a normal Move struct that cannot be used to construct Sui objects, since it doesn't have `key` ability. This is very common usage when we just need to organize data with good encapsulation.

Reviewer: All of Sam's changes LGTM, so I am committing them before beginning my review to avoid duplicate comments/commits.

In some cases, however, we want to put a Sui object struct type as a field in another Sui object struct type. In the above example, if we change `Bar` into:
```rust
struct Bar has key, store {
    id: VersionedID,
    value: u64,
}
```

> Now `Bar` is also a Sui object type. When we put a Sui object of type `Bar` into a Sui object of type `Foo`, the Sui object of type `Bar` is said to be **wrapped**.

Reviewer: Makes sense to me.

> There are some interesting consequences of wrapping an Sui object into another. When an object is wrapped, this object no longer exists independently on-chain. One will no longer be able to look up this object by its ID. This object becomes part of the data of the object that wraps it. Most importantly, one can no longer pass the wrapped object as an argument in any way in Move calls. The only access point is through the wrapping object.

Reviewer 2: Do should probably explain 3 edge cases (probably with a note):

1. It's not clear if we can wrap an already wrapped object?
wrap(wrap(wrap(obj)))

2. Are accidental cycles resolved automatically at compilation or the dev should take care of it?

3. Can a regular struct hold a Sui object (thus the opposite from the example)?

Author: Good call! Will explain them in the doc.
To answer your question
1) No. A wrapped object cannot be used directly.
2) Since 1) is no, there cannot be cycles
3) No.

```rust
public(script) fun create_object(scarcity: u8, style: u8, ctx: &mut TxContext) {
    let object = Object {
        id: TxContext::new_id(ctx),
        scarcity,
        style,
    };
    Transfer::transfer(object, TxContext::sender(ctx))
}
```
Anyone can call `create_object` to create a new object with specified `scarcity` and `style`. The created object will be sent to the signer of the transaction. We will likely also want to be able to transfer the object to others:
```rust
public(script) fun transfer_object(object: Object, ctx: &mut TxContext) {
    Transfer::transfer(object, TxContext::sender(ctx))
}
```

Now let's look at how we could enable a swap/trade between your object and others' objects. A straightforward idea is this: define a function that takes two objects from two accounts, and swap their ownership. But this doesn't work in Sui! Recall from Chapter 2 that only object owners can send a transaction to mutate the object. So one person cannot send a transaction that would swap their own object with someone else's object.

> In the future, we will likely introduce multi-sig transactions so that two people can sign the same transaction for this type of usecase. However you may not always be able to find someone to swap with right away. A multi-sig transaction won't work in this scenario. Even if you can, you may not want to carry the burden of finding a swap target.

Reviewer 3: "Multi-sig" usually refers to requiring multiple keys to send a tx from a single address rather than sending a tx that acts on behalf of multiple addresses. We used to call the idea you are describing ["multi-agent"](https://github.com/diem/dip/blob/main/dips/dip-169.md) transactions.

> Another common solution is to "send" your object to a pool (e.g. a marketplace in the case of NFT, or a liquidity pool in the case of tokens), and perform the swap in the pool (either right away, or latter when there is demand). In future chapters, we will explore the concept of shared objects that can be mutated by anyone, and show that how it enables anyone to operate in a shared object pool. In this chapter, we will focus on how to achieve the same effect using owned objects. Transactions only using owned objects is significantly faster and cheapter (in terms of gas) than using shared objects, since it does not require consensus in Sui.

Reviewer 3: ```suggestion
Another common solution is to "send" your object to a pool (e.g. a marketplace in the case of NFT, or a liquidity pool in the case of tokens), and perform the swap in the pool (either right away, or later when there is demand). In future chapters, we will explore the concept of shared objects that can be mutated by anyone, and show that how it enables anyone to operate in a shared object pool. In this chapter, we will focus on how to achieve the same effect using owned objects. Transactions only using owned objects is faster and cheaper (in terms of gas) than using shared objects, since it does not require consensus in Sui.
```

> To be able to perform a swap of objects, both objects must be owned by the same account. We can imagine that a third-party builds infrastructure to provide swap serviecs. Anyone who wants to swap their object can send their objects to the third-party, and the third-party will help perform the swap and send them back. However, we don't fully trust the third-party and don't want to give them full custody of our objects. To achieve this, we can use direct wrapping. We define a wrapper object type as following:

Reviewer 3: ```suggestion
To be able to perform a swap of objects, both objects must be owned by the same account. We can imagine that a third party builds infrastructure to provide swap services. Anyone who wants to swap their object can send their objects to the third party, and the third party will help perform the swap and send them back. However, we don't fully trust the third party and don't want to give them full custody of our objects. To achieve this, we can use direct wrapping. We define a wrapper object type as following:
```

Reviewer 3: I believe it's "third-party" when used as an adjective and "third party" when used as a noun.

Reviewer: You nailed it. That adjective being a:
https://en.wikipedia.org/wiki/Compound_modifier

```rust
public(script) fun request_swap(object: Object, fee: Coin<SUI>, service_address: address, ctx: &mut TxContext) {
    assert!(Coin::value(&fee) >= MIN_FEE, 0);
    let wrapper = ObjectWrapper {
        id: TxContext::new_id(ctx),
        original_owner: TxContext::sender(ctx),
        to_swap: object,
        fee: Coin::into_balance(fee),
    };
    Transfer::transfer(wrapper, service_address);
}
```
In the above entry function, to request for swapping an `object`, one must pass the object by value so that it's fully consumed and wrapped into `ObjectWrapper`. A fee (in the type of `Coin<SUI>`) is also provided. It also checks that the fee is sufficient. Note that we turn `Coin` into `Balance` when putting it into the `wrapper` object. This is because `Coin` is a Sui object type and only used to pass around as Sui objects (e.g. as entry function arguments or objects sent to addresses). For coin balances that need to be embedded in another Sui object struct, we use `Balance` instead because it's not a Sui object type and hence is much cheaper to use.

> The wrapper object is then sent to the service operator, whose address is also specified in the call as `serviec_address`.

Reviewer 3: ```suggestion
The wrapper object is then sent to the service operator, whose address is also specified in the call as `service_address`.
```

> Although the service operator (`service_address`) now owns the `ObjectWrapper`, which contains the object to be swapped, there is nothing they can do about the object. They cannot transfer it even though we have defined a `transfer_object` entry function for `Object`. This is because they cannot pass the wrapped `Object` as an argument to the function.

Reviewer 3: ```suggestion
Although the service operator (`service_address`) now owns the `ObjectWrapper`, which contains the object to be swapped, they cannot access the wrapped `Object. In particular, the operator cannot transfer it even though we have defined a `transfer_object` entry function for `Object`. This is because they cannot pass the wrapped `Object` as an argument to the function.
```

```rust
struct Sword has key, store {
    id: VersionedID,
    strength: u8,
}

struct Shield has key, store {
    id: VersionedID,
    armor: u8,
}
```
When we are creating a new warrior, we can set the `sword` and `shield` to none to indicate that there are no equipment yet:
```rust
public(script) fun create_warrior(ctx: &mut TxContext) {
    let warrior = SimpleWarrior {
        id: TxContext::new_id(ctx),
        sword: Option::none(),
        shield: Option::none(),
    };
    Transfer::transfer(warrior, TxContext::sender(ctx))
}
```
With this, we can then define functions to equip new swords or new shields:
```rust
public(script) fun equip_sword(warrior: &mut SimpleWarrior, sword: Sword, ctx: &mut TxContext) {
    if (Option::is_some(&warrior.sword)) {
        let old_sword = Option::extract(&mut warrior.sword);
        Transfer::transfer(old_sword, TxContext::sender(ctx));
    };
    Option::fill(&mut warrior.sword, sword);
}
```
In the above function, we are passing a `warrior` as mutable reference of `SimpleWarrior`, and a `sword` passed by value because we need to wrap it into the `warrior`.

> It is important to note that because `Sword` is a Sui object type without `drop` ability, if the warrior already has a sword equipped, that sword cannot just be dropped. If we make a call to `Option::fill` without first checking and taking out the existig sword, a runtime error may happen. Hence in `equip_sword`, we first check if there is already a sword equipped, and if so, we take it out and send it back to the sender. This matches what you would typically expect when you equip a new sword: if a sword is already equipped, you get it back.

Reviewer 3: ```suggestion
It is important to note that because `Sword` is a Sui object type without `drop` ability, if the warrior already has a sword equipped, that sword cannot just be dropped. If we make a call to `Option::fill` without first checking and taking out the existig sword, a runtime error may happen. Hence in `equip_sword`, we first check if there is already a sword equipped, and if so, we take it out and send it back to the sender. This matches what you would expect when you equip a new sword--you get the old sword back, if there is one.
```

## Chapter 4: Object Wrapping
In many programming languages, we organize data structures in layers by nesting complex data structures in another data structure. In Move, you may do the same by putting a field of `struct` type in another, like the following:
```rust
struct Foo has key {
    id: VersionedID,
    bar: Bar,
}
struct Bar has store {
    value: u64,
}
```

> > :book: For a struct type to be capable of being embedded in a Sui object struct (which will have `key` ability), the embedded struct type must have the `store` ability.

Reviewer 2: thanks!

    value: u64,
}
```
Now `Bar` is also a Sui object type. When we put a Sui object of type `Bar` into a Sui object of type `Foo`, the Sui object of type `Bar` is said to be **wrapped** by `Foo` (which we call the **wrapper** object).

> :bulb: In Move code, it is also possible to put a Sui object as a field of a non-Sui object struct type. For example, in the above code sample, we can define `Foo` to not have `key` but `Bar` to have `key, store`. However, this case can happen only temporarily in the middle of a Move execution and cannot be persisted on-chain. This is because a non-Sui object cannot flow across the Move-Sui boundary, and one must unpack the non-Sui object at some point and deal with the Sui object fields in it.

There are some interesting consequences of wrapping an Sui object into another. When an object is wrapped, this object no longer exists independently on-chain. We will no longer be able to look up this object by its ID. This object becomes part of the data of the object that wraps it. Most importantly, *we can no longer pass the wrapped object as an argument in any way in Move calls*. The only access point is through the wrapping object.

> >:bulb: The fact that you can no longer use a wrapped Sui object means that it's impossible to create circular wrapping behavior, where A wraps B, B wraps C, and C also wraps A.

Reviewer 2: <3
